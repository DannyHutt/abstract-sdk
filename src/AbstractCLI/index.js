// @flow
import path from "path";
import { spawn } from "child_process";
import { Buffer } from "buffer";
import find from "lodash/find";
import locatePath from "locate-path";
import JSONStream from "JSONStream";
import { ref, pageFileDescriptor } from "../utils";
import { log } from "../debug";
import type {
  AbstractInterface,
  ProjectDescriptor,
  CommitDescriptor,
  BranchDescriptor,
  PageDescriptor,
  FileDescriptor,
  LayerDescriptor,
  CollectionDescriptor
} from "../";

const logSpawn = log.extend("AbstractCLI:spawn");
const logError = log.extend("AbstractCLI:error");
const logStdoutError = log.extend("AbstractCLI:stdout:error");
const logStdoutData = log.extend("AbstractCLI:stdout:data");
const logStderrData = log.extend("AbstractCLI:stderr");
const logClose = log.extend("AbstractCLI:close");

function parsePath(input: ?string): ?Array<string> {
  if (!input) return;
  return input.split(path.delimiter || ":");
}

export type Options = {
  accessToken: string,
  cliPath?: string[],
  apiUrl: string,
  cwd?: string
};

export default class AbstractCLI implements AbstractInterface {
  accessToken: string;
  cliPath: string;
  apiUrl: string;
  cwd: string;

  constructor({
    cwd = process.cwd(),
    accessToken,
    cliPath = parsePath(process.env.ABSTRACT_CLI_PATH) || [
      // Relative to cwd
      path.join(cwd, "abstract-cli"),
      // Relative to node_modules in cwd (also makes test easier)
      path.join(
        cwd,
        "node_modules/@elasticprojects/abstract-cli/bin/abstract-cli"
      ),
      // macOS App
      "/Applications/Abstract.app/Contents/Resources/app.asar.unpacked/node_modules/@elasticprojects/abstract-cli/bin/abstract-cli"
    ],
    apiUrl
  }: Options) {
    this.cwd = cwd;
    this.accessToken = accessToken;
    this.apiUrl = apiUrl;

    try {
      this.cliPath = path.relative(
        cwd,
        path.resolve(cwd, locatePath.sync(cliPath))
      );
    } catch (error) {
      throw new Error(`Cannot find abstract-cli in "${cliPath.join(":")}"`);
    }
  }

  async spawn(args: string[]) {
    return new Promise((resolve, reject) => {
      let authArgs = [];

      if (this.accessToken) {
        authArgs = ["--user-token", this.accessToken];
      }

      const spawnArgs = [
        `./${path.relative(this.cwd, this.cliPath)}`,
        [
          ...authArgs,
          "--api-url",
          this.apiUrl,
          ...args // First args win for https://github.com/spf13/cobra
        ],
        { cwd: this.cwd }
      ];

      logSpawn(spawnArgs);
      const abstractCli = spawn(...spawnArgs);

      let stderrBuffer = new Buffer.from("");
      abstractCli.stderr.on("data", chunk => {
        logStderrData(chunk.toString());
        stderrBuffer = Buffer.concat([stderrBuffer, chunk]);
      });

      abstractCli.stdout
        .pipe(JSONStream.parse())
        .on("data", data => {
          logStdoutData(data);
          resolve(data);
        })
        .on("error", error => {
          logStdoutError(error.toString());
          reject(error);
        });

      abstractCli.on("error", reject);
      abstractCli.on("close", errorCode => {
        logClose(errorCode);

        if (errorCode !== 0) {
          logError(stderrBuffer.toString());
          reject(stderrBuffer); // Reject stderr for non-zero error codes
        }
      });
    });
  }

  commits = {
    list: (
      objectDescriptor: BranchDescriptor | FileDescriptor | LayerDescriptor
    ) => {
      const fileIdArgs =
        objectDescriptor.fileId && !objectDescriptor.layerId
          ? ["--file-id", objectDescriptor.fileId]
          : [];

      const layerIdArgs = objectDescriptor.layerId
        ? ["--layer-id", objectDescriptor.layerId]
        : [];

      return this.spawn([
        "commits",
        objectDescriptor.projectId,
        objectDescriptor.branchId,
        ...fileIdArgs,
        ...layerIdArgs
      ]);
    },
    info: (
      objectDescriptor: BranchDescriptor | FileDescriptor | LayerDescriptor
    ) => {
      return this.spawn([
        "commit",
        objectDescriptor.projectId,
        ref(objectDescriptor)
      ]);
    }
  };

  changesets = {
    info: async (commitDescriptor: CommitDescriptor) => {
      return this.spawn([
        "changeset",
        commitDescriptor.projectId,
        "--commit",
        commitDescriptor.sha,
        "--branch",
        commitDescriptor.branchId
      ]);
    }
  };

  files = {
    list: (branchDescriptor: BranchDescriptor) => {
      return this.spawn([
        "files",
        branchDescriptor.projectId,
        ref(branchDescriptor)
      ]);
    },
    info: (fileDescriptor: FileDescriptor) => {
      return this.spawn([
        "file",
        fileDescriptor.projectId,
        ref(fileDescriptor),
        fileDescriptor.fileId
      ]);
    }
  };

  pages = {
    list: async (fileDescriptor: FileDescriptor) => {
      return await this.files.info(fileDescriptor);
    },
    info: async (pageDescriptor: PageDescriptor) => {
      const { pages } = await this.files.info(
        pageFileDescriptor(pageDescriptor)
      );

      return find(pages, { id: pageDescriptor.pageId });
    }
  };

  layers = {
    list: (fileDescriptor: FileDescriptor) => {
      return this.spawn([
        "layers",
        fileDescriptor.projectId,
        ref(fileDescriptor),
        fileDescriptor.fileId
      ]);
    },
    info: (layerDescriptor: LayerDescriptor) => {
      return this.spawn([
        "layer",
        "meta",
        layerDescriptor.projectId,
        ref(layerDescriptor),
        layerDescriptor.fileId,
        layerDescriptor.layerId
      ]);
    }
  };

  data = {
    info: (layerDescriptor: LayerDescriptor) => {
      return this.spawn([
        "layer",
        "data",
        layerDescriptor.projectId,
        ref(layerDescriptor),
        layerDescriptor.fileId,
        layerDescriptor.layerId
      ]);
    }
  };

  collections = {
    list: (projectOrBranchDescriptor: ProjectDescriptor | BranchDescriptor) => {
      const branchArgs = projectOrBranchDescriptor.branchId
        ? ["--branch", projectOrBranchDescriptor.branchId]
        : [];

      return this.spawn([
        "collections",
        projectOrBranchDescriptor.projectId,
        ...branchArgs
      ]);
    },
    info: (collectionDescriptor: CollectionDescriptor) => {
      return this.spawn([
        "collection",
        collectionDescriptor.projectId,
        collectionDescriptor.collectionId
      ]);
    }
  };
}
