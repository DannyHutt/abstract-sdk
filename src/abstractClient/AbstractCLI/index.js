// @flow
import path from "path";
import debug from "debug";
import locatePath from "locate-path";
import type {
  AbstractTransport,
  AbstractInterface,
  ProjectDescriptor
} from "../../";

function parsePath(input: ?string): ?Array<string> {
  if (!input) return;
  return input.split(path.delimiter || ":");
}

type Options = {
  abstractToken: string,
  abstractCliPath?: string[],
  cwd?: string
};

export default class AbstractCLI
  implements AbstractTransport, AbstractInterface {
  abstractToken: string;
  abstractCliPath: string;

  constructor({
    cwd = process.cwd(),
    abstractToken,
    abstractCliPath = parsePath(process.env.ABSTRACT_CLI_PATH) || [
      path.join(cwd, "abstract-cli"), // Relative to cwd
      path.join(cwd, "node_modules/@elasticprojects/abstract-cli"), // Relative to node_modules in cwd (also makes test easier)
      "/Applications/Abstract.app/Contents/Resources/app.asar.unpacked/node_modules/@elasticprojects/abstract-cli" // macOS App
    ]
  }: Options) {
    this.abstractToken = abstractToken;

    try {
      this.abstractCliPath = path.resolve(
        cwd,
        locatePath.sync(abstractCliPath)
      );
    } catch (error) {
      throw new Error(
        `Cannot find abstract-cli in "${abstractCliPath.join(":")}"`
      );
    }
  }

  layers = {
    async data(projectDescriptor: ProjectDescriptor) {
      return await this.send(1);
    }
  };

  async send(action: Object) {
    debug("abstract:transport:AbstractCLI")(action);
    return action;
  }
}
