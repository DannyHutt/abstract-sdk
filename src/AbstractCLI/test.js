// @flow
import child_process from "child_process";
import { Readable } from "readable-stream";
import get from "lodash/get";
import {
  buildOptions,
  buildProjectDescriptor,
  buildCommitDescriptor,
  buildBranchDescriptor,
  buildFileDescriptor,
  buildPageDescriptor,
  buildLayerDescriptor,
  buildCollectionDescriptor
} from "../support/factories";
import AbstractCLI from "./";

jest.mock("child_process");

function buildTextStream(text?: string): ReadableStream {
  const stream = new Readable();
  stream._read = () => {}; // required interface

  if (text !== undefined) {
    stream.push(text);
    stream.push(null); // end of stream
  }

  return stream;
}

const responses = {
  files: {
    list: () => ({
      stdout: JSON.stringify({
        files: [{ id: "file-id" }, { id: "not-file-id" }]
      })
    }),
    info: () => ({
      stdout: JSON.stringify({ file: { id: "file-id" } })
    })
  },
  pages: {
    info: () => ({
      stdout: JSON.stringify({
        pages: [{ id: "not-page-id" }, { id: "page-id" }],
        file: {} // required
      })
    })
  }
};

describe(AbstractCLI, () => {
  test("throws when abstract-cli cannot be located", () => {
    expect(
      () =>
        new AbstractCLI(
          buildOptions({
            abstractCliPath: ["./fixtures/missing/abstract-cli"]
          })
        ).abstractCliPath
    ).toThrow('Cannot find abstract-cli in "./fixtures/missing/abstract-cli"');
  });

  describe("without module cache to reset process.env", () => {
    beforeEach(() => {
      jest.resetModules();
    });

    test.each([
      [
        // Relative path
        "./fixtures/abstract-cli",
        "fixtures/abstract-cli"
      ],
      [
        // Ignores misssing abstract-cli path
        "/missing/abstract-cli:./fixtures/abstract-cli",
        "fixtures/abstract-cli"
      ],
      [
        // Prefers first matched abstract-cli path
        "./fixtures/other/abstract-cli/:fixtures/abstract-cli",
        "fixtures/other/abstract-cli"
      ]
    ])(
      "configures options.abstractCliPath from process.env.ABSTRACT_CLI_PATH=%p",
      (abstractCliPath, expected) => {
        process.env.ABSTRACT_CLI_PATH = abstractCliPath;
        const { default: AbstractCLI } = require("./");
        expect(new AbstractCLI(buildOptions()).abstractCliPath).toBe(expected);
      }
    );

    test("prefers options.abstractCliPath=%p over process.env.ABSTRACT_CLI_PATH", () => {
      process.env.ABSTRACT_CLI_PATH = "./fixtures/other/abstract-cli";
      const { default: AbstractCLI } = require("./");
      expect(
        new AbstractCLI(
          buildOptions({
            abstractCliPath: ["./fixtures/abstract-cli"]
          })
        ).abstractCliPath
      ).toBe("fixtures/abstract-cli");
    });
  });

  describe("with mocked child_process.spawn", () => {
    beforeEach(() => {
      jest.resetAllMocks();
    });

    test.each([
      // collections
      ["collections.list", buildProjectDescriptor()],
      ["collections.list", buildBranchDescriptor()],
      ["collections.info", buildCollectionDescriptor()],
      // commits
      ["commits.list", buildBranchDescriptor()],
      ["commits.list", buildFileDescriptor()],
      ["commits.list", buildLayerDescriptor()],
      ["commits.list", buildBranchDescriptor({ sha: "sha" })],
      ["commits.list", buildFileDescriptor({ sha: "sha" })],
      ["commits.list", buildLayerDescriptor({ sha: "sha" })],
      ["commits.info", buildBranchDescriptor()],
      ["commits.info", buildFileDescriptor()],
      ["commits.info", buildLayerDescriptor()],
      ["commits.info", buildBranchDescriptor({ sha: "sha" })],
      ["commits.info", buildFileDescriptor({ sha: "sha" })],
      ["commits.info", buildLayerDescriptor({ sha: "sha" })],
      // changesets
      ["changesets.info", buildCommitDescriptor()],
      // files
      [
        "files.info",
        buildFileDescriptor(),
        {
          responses: [responses.files.info()]
        }
      ],
      [
        "files.list",
        buildBranchDescriptor({ sha: "sha" }),
        {
          responses: [responses.files.list()]
        }
      ],
      [
        "files.list",
        buildCommitDescriptor({ sha: "sha" }),
        {
          responses: [responses.files.list()]
        }
      ],
      [
        "files.info",
        buildFileDescriptor({ sha: "sha" }),
        {
          responses: [responses.files.info()]
        }
      ],
      // pages
      ["pages.list", buildFileDescriptor()],
      ["pages.list", buildFileDescriptor({ sha: "sha" })],
      [
        "pages.info",
        buildPageDescriptor(),
        {
          responses: [responses.pages.info()],
          result: { id: "page-id" }
        }
      ],
      [
        "pages.info",
        buildPageDescriptor({ sha: "sha" }),
        {
          responses: [responses.pages.info()],
          result: { id: "page-id" }
        }
      ],
      // layers
      ["layers.list", buildFileDescriptor()],
      ["layers.info", buildLayerDescriptor()],
      ["layers.list", buildFileDescriptor({ sha: "sha" })],
      ["layers.info", buildLayerDescriptor({ sha: "sha" })],
      // data
      ["data.info", buildLayerDescriptor()],
      ["data.info", buildLayerDescriptor({ sha: "sha" })]
    ])("%s(%p)", async (property, args, options = {}) => {
      args = Array.isArray(args) ? args : [args];

      const transport = new AbstractCLI(
        buildOptions({ abstractCliPath: ["./fixtures/abstract-cli"] })
      );

      const transportMethod = get(transport, property).bind(transport);

      if (options.responses) {
        options.responses.forEach(response => {
          child_process.spawn.mockReturnValueOnce({
            stdout: buildTextStream(response.stdout),
            stderr: buildTextStream(response.stderr),
            on: jest.fn().mockReturnThis()
          });
        });
      }

      child_process.spawn.mockReturnValueOnce({
        stdout: buildTextStream("{}"),
        stderr: buildTextStream(""),
        on: jest.fn().mockReturnThis()
      });

      const result = transportMethod(...args);
      await expect(result).resolves;

      if (options.result) {
        expect(await result).toEqual(options.result);
      }

      expect(child_process.spawn.mock.calls.length).toEqual(1);
      expect(child_process.spawn.mock.calls[0]).toMatchSnapshot();
    });
  });
});
