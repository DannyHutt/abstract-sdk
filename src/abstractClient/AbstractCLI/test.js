// @flow
import path from "path";
import get from "lodash/get";
import AbstractCLI from "./";

const EXAMPLE_OPTIONS = { abstractToken: "1" };

describe(AbstractCLI, () => {
  test("throws when abstract-cli cannot be located", () => {
    expect(
      () =>
        new AbstractCLI({
          ...EXAMPLE_OPTIONS,
          abstractCliPath: ["./fixtures/missing/abstract-cli"]
        }).abstractCliPath
    ).toThrow('Cannot find abstract-cli in "./fixtures/missing/abstract-cli"');
  });

  describe("without module cache to reset process.env", () => {
    beforeEach(() => {
      jest.resetModules();
    });

    test.each([
      [
        // Relative path
        "./fixtures/abstract-cli",
        path.join(process.cwd(), "./fixtures/abstract-cli")
      ],
      [
        // Ignores misssing abstract-cli path
        "/missing/abstract-cli:./fixtures/abstract-cli",
        path.join(process.cwd(), "./fixtures/abstract-cli")
      ],
      [
        // Prefers first matched abstract-cli path
        "./fixtures/other/abstract-cli/:fixtures/abstract-cli",
        path.join(process.cwd(), "./fixtures/other/abstract-cli")
      ]
    ])(
      "configures options.abstractCliPath from process.env.ABSTRACT_CLI_PATH=%p",
      (abstractCliPath, expected) => {
        process.env.ABSTRACT_CLI_PATH = abstractCliPath;
        const { default: AbstractCLI } = require("./");
        expect(new AbstractCLI(EXAMPLE_OPTIONS).abstractCliPath).toBe(expected);
      }
    );

    test("prefers options.abstractCliPath=%p over process.env.ABSTRACT_CLI_PATH", () => {
      process.env.ABSTRACT_CLI_PATH = "./fixtures/other/abstract-cli";
      const { default: AbstractCLI } = require("./");
      expect(
        new AbstractCLI({
          ...EXAMPLE_OPTIONS,
          abstractCliPath: ["./fixtures/abstract-cli"]
        }).abstractCliPath
      ).toBe(path.join(process.cwd(), "./fixtures/abstract-cli"));
    });
  });

  test.each([
    ["layers.data", {}, 1],
    ["layers.data", {}, 1],
    ["layers.data", {}, 1],
    ["layers.data", {}, 1]
  ])("%s(%p): Promise<%p>", async (property, descriptor, expected) => {
    const transport = new AbstractCLI(EXAMPLE_OPTIONS);
    const transportMethod = get(transport, property).bind(transport);
    await expect(transportMethod(descriptor)).resolves.toBe(expected);
  });
});
