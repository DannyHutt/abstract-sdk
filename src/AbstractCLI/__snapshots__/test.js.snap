// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AbstractCLI with mocked child_process.spawn changesets.info({"branchId": "branch-id", "projectId": "project-id", "sha": "commit-sha"}) 1`] = `
Array [
  "./fixtures/abstract-cli",
  Array [
    "--user-token",
    "abstract-token",
    "--api-url",
    "https://api.goabstract.com",
    "changeset",
    "project-id",
    "--commit",
    "commit-sha",
    "--branch",
    "branch-id",
  ],
  Object {
    "cwd": "<PROJECT_ROOT>",
  },
]
`;

exports[`AbstractCLI with mocked child_process.spawn collections.info({"branchId": "branch-id", "collectionId": "collection-id", "projectId": "project-id"}) 1`] = `
Array [
  "./fixtures/abstract-cli",
  Array [
    "--user-token",
    "abstract-token",
    "--api-url",
    "https://api.goabstract.com",
    "collection",
    "project-id",
    "collection-id",
  ],
  Object {
    "cwd": "<PROJECT_ROOT>",
  },
]
`;

exports[`AbstractCLI with mocked child_process.spawn collections.list({"branchId": "branch-id", "projectId": "project-id"}) 1`] = `
Array [
  "./fixtures/abstract-cli",
  Array [
    "--user-token",
    "abstract-token",
    "--api-url",
    "https://api.goabstract.com",
    "collections",
    "project-id",
    "--branch",
    "branch-id",
  ],
  Object {
    "cwd": "<PROJECT_ROOT>",
  },
]
`;

exports[`AbstractCLI with mocked child_process.spawn collections.list({"projectId": "project-id"}) 1`] = `
Array [
  "./fixtures/abstract-cli",
  Array [
    "--user-token",
    "abstract-token",
    "--api-url",
    "https://api.goabstract.com",
    "collections",
    "project-id",
  ],
  Object {
    "cwd": "<PROJECT_ROOT>",
  },
]
`;

exports[`AbstractCLI with mocked child_process.spawn commits.info({"branchId": "branch-id", "fileId": "file-id", "layerId": "layer-id", "pageId": "page-id", "projectId": "project-id", "sha": "layer-sha"}) 1`] = `
Array [
  "./fixtures/abstract-cli",
  Array [
    "--user-token",
    "abstract-token",
    "--api-url",
    "https://api.goabstract.com",
    "commit",
    "project-id",
    "layer-sha",
  ],
  Object {
    "cwd": "<PROJECT_ROOT>",
  },
]
`;

exports[`AbstractCLI with mocked child_process.spawn commits.info({"branchId": "branch-id", "fileId": "file-id", "layerId": "layer-id", "pageId": "page-id", "projectId": "project-id", "sha": "sha"}) 1`] = `
Array [
  "./fixtures/abstract-cli",
  Array [
    "--user-token",
    "abstract-token",
    "--api-url",
    "https://api.goabstract.com",
    "commit",
    "project-id",
    "sha",
  ],
  Object {
    "cwd": "<PROJECT_ROOT>",
  },
]
`;

exports[`AbstractCLI with mocked child_process.spawn commits.info({"branchId": "branch-id", "fileId": "file-id", "projectId": "project-id", "sha": "sha"}) 1`] = `
Array [
  "./fixtures/abstract-cli",
  Array [
    "--user-token",
    "abstract-token",
    "--api-url",
    "https://api.goabstract.com",
    "commit",
    "project-id",
    "sha",
  ],
  Object {
    "cwd": "<PROJECT_ROOT>",
  },
]
`;

exports[`AbstractCLI with mocked child_process.spawn commits.info({"branchId": "branch-id", "fileId": "file-id", "projectId": "project-id"}) 1`] = `
Array [
  "./fixtures/abstract-cli",
  Array [
    "--user-token",
    "abstract-token",
    "--api-url",
    "https://api.goabstract.com",
    "commit",
    "project-id",
    "branch-id",
  ],
  Object {
    "cwd": "<PROJECT_ROOT>",
  },
]
`;

exports[`AbstractCLI with mocked child_process.spawn commits.info({"branchId": "branch-id", "projectId": "project-id", "sha": "sha"}) 1`] = `
Array [
  "./fixtures/abstract-cli",
  Array [
    "--user-token",
    "abstract-token",
    "--api-url",
    "https://api.goabstract.com",
    "commit",
    "project-id",
    "sha",
  ],
  Object {
    "cwd": "<PROJECT_ROOT>",
  },
]
`;

exports[`AbstractCLI with mocked child_process.spawn commits.info({"branchId": "branch-id", "projectId": "project-id"}) 1`] = `
Array [
  "./fixtures/abstract-cli",
  Array [
    "--user-token",
    "abstract-token",
    "--api-url",
    "https://api.goabstract.com",
    "commit",
    "project-id",
    "branch-id",
  ],
  Object {
    "cwd": "<PROJECT_ROOT>",
  },
]
`;

exports[`AbstractCLI with mocked child_process.spawn commits.list({"branchId": "branch-id", "fileId": "file-id", "layerId": "layer-id", "pageId": "page-id", "projectId": "project-id", "sha": "layer-sha"}) 1`] = `
Array [
  "./fixtures/abstract-cli",
  Array [
    "--user-token",
    "abstract-token",
    "--api-url",
    "https://api.goabstract.com",
    "commits",
    "project-id",
    "branch-id",
    "--layer-id",
    "layer-id",
  ],
  Object {
    "cwd": "<PROJECT_ROOT>",
  },
]
`;

exports[`AbstractCLI with mocked child_process.spawn commits.list({"branchId": "branch-id", "fileId": "file-id", "layerId": "layer-id", "pageId": "page-id", "projectId": "project-id", "sha": "sha"}) 1`] = `
Array [
  "./fixtures/abstract-cli",
  Array [
    "--user-token",
    "abstract-token",
    "--api-url",
    "https://api.goabstract.com",
    "commits",
    "project-id",
    "branch-id",
    "--layer-id",
    "layer-id",
  ],
  Object {
    "cwd": "<PROJECT_ROOT>",
  },
]
`;

exports[`AbstractCLI with mocked child_process.spawn commits.list({"branchId": "branch-id", "fileId": "file-id", "projectId": "project-id", "sha": "sha"}) 1`] = `
Array [
  "./fixtures/abstract-cli",
  Array [
    "--user-token",
    "abstract-token",
    "--api-url",
    "https://api.goabstract.com",
    "commits",
    "project-id",
    "branch-id",
    "--file-id",
    "file-id",
  ],
  Object {
    "cwd": "<PROJECT_ROOT>",
  },
]
`;

exports[`AbstractCLI with mocked child_process.spawn commits.list({"branchId": "branch-id", "fileId": "file-id", "projectId": "project-id"}) 1`] = `
Array [
  "./fixtures/abstract-cli",
  Array [
    "--user-token",
    "abstract-token",
    "--api-url",
    "https://api.goabstract.com",
    "commits",
    "project-id",
    "branch-id",
    "--file-id",
    "file-id",
  ],
  Object {
    "cwd": "<PROJECT_ROOT>",
  },
]
`;

exports[`AbstractCLI with mocked child_process.spawn commits.list({"branchId": "branch-id", "projectId": "project-id", "sha": "sha"}) 1`] = `
Array [
  "./fixtures/abstract-cli",
  Array [
    "--user-token",
    "abstract-token",
    "--api-url",
    "https://api.goabstract.com",
    "commits",
    "project-id",
    "branch-id",
  ],
  Object {
    "cwd": "<PROJECT_ROOT>",
  },
]
`;

exports[`AbstractCLI with mocked child_process.spawn commits.list({"branchId": "branch-id", "projectId": "project-id"}) 1`] = `
Array [
  "./fixtures/abstract-cli",
  Array [
    "--user-token",
    "abstract-token",
    "--api-url",
    "https://api.goabstract.com",
    "commits",
    "project-id",
    "branch-id",
  ],
  Object {
    "cwd": "<PROJECT_ROOT>",
  },
]
`;

exports[`AbstractCLI with mocked child_process.spawn data.layer({"branchId": "branch-id", "fileId": "file-id", "layerId": "layer-id", "pageId": "page-id", "projectId": "project-id", "sha": "layer-sha"}) 1`] = `
Array [
  "./fixtures/abstract-cli",
  Array [
    "--user-token",
    "abstract-token",
    "--api-url",
    "https://api.goabstract.com",
    "layer",
    "data",
    "project-id",
    "layer-sha",
    "file-id",
    "layer-id",
  ],
  Object {
    "cwd": "<PROJECT_ROOT>",
  },
]
`;

exports[`AbstractCLI with mocked child_process.spawn data.layer({"branchId": "branch-id", "fileId": "file-id", "layerId": "layer-id", "pageId": "page-id", "projectId": "project-id", "sha": "sha"}) 1`] = `
Array [
  "./fixtures/abstract-cli",
  Array [
    "--user-token",
    "abstract-token",
    "--api-url",
    "https://api.goabstract.com",
    "layer",
    "data",
    "project-id",
    "sha",
    "file-id",
    "layer-id",
  ],
  Object {
    "cwd": "<PROJECT_ROOT>",
  },
]
`;

exports[`AbstractCLI with mocked child_process.spawn files.info({"branchId": "branch-id", "fileId": "file-id", "projectId": "project-id", "sha": "sha"}) 1`] = `
Array [
  "./fixtures/abstract-cli",
  Array [
    "--user-token",
    "abstract-token",
    "--api-url",
    "https://api.goabstract.com",
    "file",
    "project-id",
    "sha",
    "file-id",
  ],
  Object {
    "cwd": "<PROJECT_ROOT>",
  },
]
`;

exports[`AbstractCLI with mocked child_process.spawn files.info({"branchId": "branch-id", "fileId": "file-id", "projectId": "project-id"}) 1`] = `
Array [
  "./fixtures/abstract-cli",
  Array [
    "--user-token",
    "abstract-token",
    "--api-url",
    "https://api.goabstract.com",
    "file",
    "project-id",
    "branch-id",
    "file-id",
  ],
  Object {
    "cwd": "<PROJECT_ROOT>",
  },
]
`;

exports[`AbstractCLI with mocked child_process.spawn files.list({"branchId": "branch-id", "projectId": "project-id", "sha": "commit-sha"}) 1`] = `
Array [
  "./fixtures/abstract-cli",
  Array [
    "--user-token",
    "abstract-token",
    "--api-url",
    "https://api.goabstract.com",
    "files",
    "project-id",
    "commit-sha",
  ],
  Object {
    "cwd": "<PROJECT_ROOT>",
  },
]
`;

exports[`AbstractCLI with mocked child_process.spawn files.list({"branchId": "branch-id", "projectId": "project-id", "sha": "sha"}) 1`] = `
Array [
  "./fixtures/abstract-cli",
  Array [
    "--user-token",
    "abstract-token",
    "--api-url",
    "https://api.goabstract.com",
    "files",
    "project-id",
    "sha",
  ],
  Object {
    "cwd": "<PROJECT_ROOT>",
  },
]
`;

exports[`AbstractCLI with mocked child_process.spawn files.list({"branchId": "branch-id", "projectId": "project-id", "sha": "sha"}) 2`] = `
Array [
  "./fixtures/abstract-cli",
  Array [
    "--user-token",
    "abstract-token",
    "--api-url",
    "https://api.goabstract.com",
    "files",
    "project-id",
    "sha",
  ],
  Object {
    "cwd": "<PROJECT_ROOT>",
  },
]
`;

exports[`AbstractCLI with mocked child_process.spawn files.list({"branchId": "branch-id", "projectId": "project-id"}) 1`] = `
Array [
  "./fixtures/abstract-cli",
  Array [
    "--user-token",
    "abstract-token",
    "--api-url",
    "https://api.goabstract.com",
    "files",
    "project-id",
    "branch-id",
  ],
  Object {
    "cwd": "<PROJECT_ROOT>",
  },
]
`;

exports[`AbstractCLI with mocked child_process.spawn layers.info({"branchId": "branch-id", "fileId": "file-id", "layerId": "layer-id", "pageId": "page-id", "projectId": "project-id", "sha": "layer-sha"}) 1`] = `
Array [
  "./fixtures/abstract-cli",
  Array [
    "--user-token",
    "abstract-token",
    "--api-url",
    "https://api.goabstract.com",
    "layer",
    "meta",
    "project-id",
    "layer-sha",
    "file-id",
    "layer-id",
  ],
  Object {
    "cwd": "<PROJECT_ROOT>",
  },
]
`;

exports[`AbstractCLI with mocked child_process.spawn layers.info({"branchId": "branch-id", "fileId": "file-id", "layerId": "layer-id", "pageId": "page-id", "projectId": "project-id", "sha": "sha"}) 1`] = `
Array [
  "./fixtures/abstract-cli",
  Array [
    "--user-token",
    "abstract-token",
    "--api-url",
    "https://api.goabstract.com",
    "layer",
    "meta",
    "project-id",
    "sha",
    "file-id",
    "layer-id",
  ],
  Object {
    "cwd": "<PROJECT_ROOT>",
  },
]
`;

exports[`AbstractCLI with mocked child_process.spawn layers.list({"branchId": "branch-id", "fileId": "file-id", "projectId": "project-id", "sha": "sha"}) 1`] = `
Array [
  "./fixtures/abstract-cli",
  Array [
    "--user-token",
    "abstract-token",
    "--api-url",
    "https://api.goabstract.com",
    "layers",
    "project-id",
    "sha",
    "file-id",
  ],
  Object {
    "cwd": "<PROJECT_ROOT>",
  },
]
`;

exports[`AbstractCLI with mocked child_process.spawn layers.list({"branchId": "branch-id", "fileId": "file-id", "projectId": "project-id"}) 1`] = `
Array [
  "./fixtures/abstract-cli",
  Array [
    "--user-token",
    "abstract-token",
    "--api-url",
    "https://api.goabstract.com",
    "layers",
    "project-id",
    "branch-id",
    "file-id",
  ],
  Object {
    "cwd": "<PROJECT_ROOT>",
  },
]
`;

exports[`AbstractCLI with mocked child_process.spawn pages.info({"branchId": "branch-id", "fileId": "file-id", "pageId": "page-id", "projectId": "project-id", "sha": "sha"}) 1`] = `
Array [
  "./fixtures/abstract-cli",
  Array [
    "--user-token",
    "abstract-token",
    "--api-url",
    "https://api.goabstract.com",
    "file",
    "project-id",
    "sha",
    "file-id",
  ],
  Object {
    "cwd": "<PROJECT_ROOT>",
  },
]
`;

exports[`AbstractCLI with mocked child_process.spawn pages.info({"branchId": "branch-id", "fileId": "file-id", "pageId": "page-id", "projectId": "project-id"}) 1`] = `
Array [
  "./fixtures/abstract-cli",
  Array [
    "--user-token",
    "abstract-token",
    "--api-url",
    "https://api.goabstract.com",
    "file",
    "project-id",
    "branch-id",
    "file-id",
  ],
  Object {
    "cwd": "<PROJECT_ROOT>",
  },
]
`;

exports[`AbstractCLI with mocked child_process.spawn pages.list({"branchId": "branch-id", "fileId": "file-id", "projectId": "project-id", "sha": "sha"}) 1`] = `
Array [
  "./fixtures/abstract-cli",
  Array [
    "--user-token",
    "abstract-token",
    "--api-url",
    "https://api.goabstract.com",
    "file",
    "project-id",
    "sha",
    "file-id",
  ],
  Object {
    "cwd": "<PROJECT_ROOT>",
  },
]
`;

exports[`AbstractCLI with mocked child_process.spawn pages.list({"branchId": "branch-id", "fileId": "file-id", "projectId": "project-id"}) 1`] = `
Array [
  "./fixtures/abstract-cli",
  Array [
    "--user-token",
    "abstract-token",
    "--api-url",
    "https://api.goabstract.com",
    "file",
    "project-id",
    "branch-id",
    "file-id",
  ],
  Object {
    "cwd": "<PROJECT_ROOT>",
  },
]
`;
